core/src/main/java/org/elasticsearch/common/path/PathTrie.java

```diff
@@ -136,23 +201,58 @@ private boolean isNamedWildcard() {
             return namedWildcard != null;
         }
 
-        public T retrieve(String[] path, int index, Map<String, String> params) {
+        public T retrieve(String[] path, int index, Map<String, String> params, TrieMatchingMode trieMatchingMode) {
             if (index >= path.length)
                 return null;
 
             String token = path[index];
             TrieNode node = children.get(token);
             boolean usedWildcard;
+
             if (node == null) {
-                node = children.get(wildcard);
-                if (node == null) {
+                if (trieMatchingMode == TrieMatchingMode.WILDCARD_NODES_ALLOWED) {
```

Comment:
I tried to change this to a `switch`, but it's not strictly a switch due to the additional `if` conditions in the branches below and the else condition, I think it's much cleaner as an `if` tree